lib/package-manager.ts

```diff
@@ -0,0 +1,56 @@
+
+import { exported } from './common/decorators';
+export class PackageManager implements INodePackageManager {
+	private packageManager: INodePackageManager;
+
+	constructor(
+		private $errors: IErrors,
+		private $npm: INodePackageManager,
+		private $options: IOptions,
+		private $yarn: INodePackageManager,
+		private $userSettingsService: IUserSettingsService
+	) {
+		this._determinePackageManager();
```

Comment:
This method is `async` so it is possible to call some method and `this.packageManager` to be null.
So I suggest an approach similar to this https://github.com/NativeScript/nativescript-cli/blob/master/lib/common/mobile/android/android-debug-bridge.ts#L56.

```
@cache()
private async init(): Promise<void> {
    this.packageManager = await this._determinePackageManager();
}

@exported("packageManager")
@invokeInit()
public install(packageName: string, pathToSave: string, config: INodePackageManagerInstallOptions): Promise<INpmInstallResultInfo> {
    return this.packageManager.install(packageName, pathToSave, config);
}
```

This way `this._determinePackageManager();` should be removed from constructor and all methods that relied on `this.packageManager` should be decorated with `@invokeInit()`

On the other side `init()` method is decorated with `cache()` decorator so the value will be persisted and `this._determinePackageManager()` will be called only once.

Also we can rewrite `_determinePackageManager` method:
```
private async _determinePackageManager(): Promise<void> {
    let pm = null;
    try {
        pm = await this.$userSettingsService.getSettingValue('packageManager');
    } catch (err) {
        this.$errors.fail(`Unable to read package manager config from user settings ${err}`);
    }

    if (pm === 'yarn' || this.$options.yarn) {
	this.packageManager = this.$yarn;
    } else {
	this.packageManager = this.$npm;
    }	
}
```